L1Trigger/TrackerTFP/src/GeometricProcessor.cc

```diff
@@ -0,0 +1,128 @@
+    enableTruncation_(iConfig.getParameter<bool>("EnableTruncation")),
+    setup_(setup),
+    dataFormats_(dataFormats),
+    region_(region),
+    input_(dataFormats_->numChannel(Process::gp), vector<deque<StubPP*>>(dataFormats_->numChannel(Process::pp))) {}
+
+  void GeometricProcessor::consume(const TTDTC& ttDTC) {
+    auto validFrame = [](int& sum, const TTDTC::Frame& frame){ return sum += frame.first.isNonnull() ? 1 : 0; };
+    int nStubsPP(0);
+    for (int channel = 0; channel < dataFormats_->numChannel(Process::pp); channel++) {
+      const TTDTC::Stream& stream = ttDTC.stream(region_, channel);
+      nStubsPP += accumulate(stream.begin(), stream.end(), 0, validFrame);
+    }
+    stubsPP_.reserve(nStubsPP);
+    for (int channel = 0; channel < dataFormats_->numChannel(Process::pp); channel++) {
+      for (const TTDTC::Frame& frame : ttDTC.stream(region_, channel)) {
+        StubPP* stub = nullptr;
+        if (frame.first.isNonnull()) {
+          stubsPP_.emplace_back(frame, dataFormats_);
+          stub = &stubsPP_.back();
+        }
+        for (int sector = 0; sector < dataFormats_->numChannel(Process::gp); sector++)
+          input_[sector][channel].push_back(stub && stub->inSector(sector) ? stub : nullptr);
+      }
+    }
+    // remove all gaps between end and last stub
+    for(vector<deque<StubPP*>>& input : input_)
+      for(deque<StubPP*>& stubs : input)
+        for(auto it = stubs.end(); it != stubs.begin();)
+          it = (*--it) ? stubs.begin() : stubs.erase(it);
+    auto validStub = [](int& sum, StubPP* stub){ return sum += stub ? 1 : 0; };
+    int nStubsGP(0);
+    for (const vector<deque<StubPP*>>& sector : input_)
+      for (const deque<StubPP*>& channel : sector)
+        nStubsGP += accumulate(channel.begin(), channel.end(), 0, validStub);
+    stubsGP_.reserve(nStubsGP);
+  }
+
+  void GeometricProcessor::produce(TTDTC::Streams& accepted, TTDTC::Streams& lost) {
+    for (int sector = 0; sector < dataFormats_->numChannel(Process::gp); sector++) {
+      vector<deque<StubPP*>>& inputs = input_[sector];
+      vector<deque<StubGP*>> stacks(dataFormats_->numChannel(Process::pp));
+      const int sectorPhi = sector % setup_->numSectorsPhi();
+      const int sectorEta = sector / setup_->numSectorsPhi();
+      auto size =  [](int& sum, const deque<StubPP*>& stubs){ return sum += stubs.size(); };
+      const int nStubs = accumulate(inputs.begin(), inputs.end(), 0, size);
+      vector<StubGP*> acceptedSector;
+      vector<StubGP*> lostSector;
+      acceptedSector.reserve(nStubs);
+      lostSector.reserve(nStubs);
+      // clock accurate firmware emulation, each while trip describes one clock tick, one stub in and one stub out per tick
+      while(!all_of(inputs.begin(), inputs.end(), [](const deque<StubPP*>& stubs){ return stubs.empty(); }) or
+            !all_of(stacks.begin(), stacks.end(), [](const deque<StubGP*>& stubs){ return stubs.empty(); })) {
+        // fill input fifo
```

Comment:
To aid comprehension add comment explaining order stubs sent to output for each sector. As far as I can tell, stubs from the 1st PP input stream always have priority over stubs from the 2nd or 3rd PP streams, several PP streams all want their stubs sent to output?

Reply:
done